lib/docker.ts

```diff
@@ -86,4 +87,85 @@ class Docker {
       throw error;
     }
   }
+
+  public async save(callback: SaveImageCallback): Promise<any> {
+    const tmpobj = fileSync({
+      mode: 0o644,
+      prefix: "docker-",
+      postfix: ".image",
+      detachDescriptor: true,
+    });
+    let err = "";
+
+    try {
+      await execute("docker", [
+        ...this.optionsList,
+        "save",
+        "-o",
+        tmpobj.name,
+        this.targetImage,
+      ]);
+    } catch (error) {
+      const stderr: string = error.stderr;
+      if (typeof stderr === "string") {
+        if (stderr.indexOf("No such image") >= 0) {
+          err = `No such image: ${this.targetImage}`;
+        } else {
+          err = error;
+        }
+      }
+    }
+
+    if (callback) {
+      try {
+        return callback(err, tmpobj.name);
+      } finally {
+        // We don't need the file anymore and could manually call the removeCallback
+        tmpobj.removeCallback();
+      }
+    }
+    // if we didn't pass the keep option the file will be deleted on exit
+  }
+
+  /**
+   * Saves the docker image as a TAR file to a temporary location and extract
+   * the specified files from it.
+   * @param paths list of files paths to extract from the image
+   */
+  public async extract(paths: string[]): Promise<{ [key: string]: Buffer }> {
+    const result: { [key: string]: Buffer } = {};
+    return this.save(async (err, name) => {
+      if (err) {
+        throw err;
+      }
+
+      const extracted: Extracted = await extractFiles(name, paths);
+
+      // No manifest file
+      if (!extracted.Manifest) {
+        return {};
+      }
+
+      const manifest = JSON.parse(extracted.Manifest);
+      const layersNames = manifest[0].Layers as string[];
+
+      if (extracted.Layers) {
+        // reverse layer order from last to first
+        for (const layerName of layersNames.reverse()) {
+          // files found for this layer
+          if (layerName in extracted.Layers) {
+            // go over files found in this layer
+            for (const filename of Object.keys(extracted.Layers[layerName])) {
+              // file was not found in previous layer
+              if (!(filename in result)) {
```

Comment:
```if (!Reflect.has(result, filename))```

Given the range of filenames possible, we need to be careful about overriding object, weirdly enough

Reply:
Thanks